src/otx/engine/utils/auto_configurator.py

```diff
@@ -0,0 +1,266 @@
+    "ava": [OTXTaskType.ACTION_DETECTION],
+}
+
+
+def configure_task(data_root: str | Path | None) -> OTXTaskType | None:
+    """Configures the task based on the given data root.
+
+    Args:
+        data_root (str | Path | None): The root directory of the data.
+
+    Returns:
+        OTXTaskType | None: The configured task type, or None if data_root is None.
+
+    Raises:
+        ValueError: If the data format is not supported.
+    """
+    if data_root is None:
+        return None
+    data_root = Path(data_root).resolve()
+
+    data_format = datumaro.Environment().detect_dataset(str(data_root))
+    if len(data_format) > 1:
+        logger.warning(f"Found multiple data formats: {data_format}. We will use the first one.")
+    data_format = data_format[0]
+    if data_format not in TASK_PER_DATA_FORMAT:
+        msg = f"Can't find proper task. we are not support {data_format} format, yet."
+        raise ValueError(msg)
+    if len(TASK_PER_DATA_FORMAT[data_format]) > 1:
+        logger.warning(
+            f"Found multiple tasks with {data_format}: {TASK_PER_DATA_FORMAT[data_format]}. We will use the first one.",
+        )
+    return TASK_PER_DATA_FORMAT[data_format][0]
+
+
+def get_num_classes_from_meta_info(task: OTXTaskType, meta_info: LabelInfo) -> int:
+    """Get the number of classes from the meta information.
+
+    Args:
+        task (OTXTaskType): The current task type.
+        meta_info (LabelInfo): The meta information about the labels.
+
+    Returns:
+        int: The number of classes.
+    """
+    num_classes = len(meta_info.label_names)
+    # Check background class
+    if task in (OTXTaskType.SEMANTIC_SEGMENTATION):
+        has_background = False
+        for label in meta_info.label_names:
+            if label.lower() == "background":
+                has_background = True
+                break
+        if not has_background:
+            num_classes += 1
+    return num_classes
+
+
+class AutoConfigurator:
+    """This Class is used to configure the OTXDataModule, OTXModel, Optimizer, and Scheduler with OTX Default.
+
+    Args:
+        data_root (str | Path | None, optional): The root directory for data storage. Defaults to None.
+        task (OTXTaskType | None, optional): The current task. Defaults to None.
+        model_name (str | None, optional): Name of the model to use as the default.
+            If None, the default model will be used. Defaults to None.
+
+    Example:
+        The following examples show how to use the AutoConfigurator class.
+
+        >>> auto_configurator = AutoConfigurator(
+        ...     data_root=<dataset/path>,
+        ...     task=<OTXTaskType>,
+        ... )
+
+        # If task is None, the task will be configured based on the data root.
+        >>> auto_configurator = AutoConfigurator(
+        ...     data_root=<dataset/path>,
+        ... )
+    """
+
+    def __init__(
+        self,
+        data_root: str | Path | None = None,
```

Comment:
Why is it `None`able?

Reply:
Can we use this type alias for the better clean code?
https://stackoverflow.com/a/77196128

Reply:
Thanks! Done - https://github.com/openvinotoolkit/training_extensions/pull/2841/commits/f861a7fb4598c0f27b3c7663f5a9cff73c85656f

Reply:
I still don't get the point why it can be `None`.

Reply:
This can be `None` in the case of a `datamodule` coming in instead of `data_root` in Engine. This does not require a `datamodule` to be provided (Since don't have a data_root, it can't create a datamodule.), but may provide a model, optimizer, or scheduler depending on the task. In this case, we don't need a `data_root` to provide a model, optimizer, and scheduler for the task.  (data_root is None, but task is not None case)